### ISSUE TITLE
MIDIFunc msgNum doesn't support the full matchItem interface

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#5103]
## Environment

* SuperCollider version: develop branch (probably 3.11 as well)
* Operating system: Linux (doesn't matter)

## Steps to reproduce

```supercollider
MIDIClient.init;

MIDIdef.noteOn(\test, { |... args| args.postln }, 49);
MIDIIn.doNoteOnAction(MIDIClient.sources[0].uid, 0, 49, 64);

// ^^ note matches, prints args

MIDIdef.noteOn(\test, { |... args| args.postln }, { |note| note.debug <= 49 });
```

At this point you get a `<=` not understood error.

If you fix that, it still doesn't match.

## Expected vs. actual behavior

`msgNum` should support the full matchItem interface (as implemented in MIDIEvent:match).

Problem 1: https://github.com/supercollider/supercollider/blob/develop/SCClassLibrary/Common/Control/ResponseDefs.sc#L813-L814:

```
		msgNum = msgNum.isNumber.if({ msgNum.asInteger }, msgNum);
		msgNum = msgNum.isCollection.if({ msgNum.collect(_.asInteger) }, msgNum);
```

Non-standard `if` syntax here disallows msgFunc to be a Function. This is why the `<=` bombs in the example. (Side comment: This is a good reason to encourage the style guide's recommendations. The above looks innocent but it is rather dangerous actually.)

Problem 2: Changing both of those to `if(condition) { true } { false }`, it still doesn't work: note number 49 is indeed <= 49, but it doesn't fire. I don't know why, and I don't have time right this second to figure out why.

So I'll assign this to myself to check later.

**Comment by muellmusik** [comment]
James apologies if I’ve misunderstood here but why does the non-standard if syntax cause the problem? From the early days of SC if(object, ...) should always be equivalent to object.if(...). This doesn’t just apply to if of course but any variation between receiver and function call syntax. So this issue aside, shouldn’t if be fixed if it’s causing this problem?

**Comment by elgiano** [comment]
Yeah, it's deeper:
- msgNum is used as a key to store the responder in the dispatcher. 
https://github.com/supercollider/supercollider/blob/435ab9106a4abdb2e3d5c076dc90a9ea9a9fd75e/SCClassLibrary/Common/Control/ResponseDefs.sc#L132
Here `key` is from `this.getKeysForFuncProxy(funcProxy)`, which for MIDIMessageDispatcher is:
https://github.com/supercollider/supercollider/blob/435ab9106a4abdb2e3d5c076dc90a9ea9a9fd75e/SCClassLibrary/Common/Control/ResponseDefs.sc#L412

- when MIDIIn has an event to dispatch, it calls .value on the dispatcher, which finds a responder by key:
https://github.com/supercollider/supercollider/blob/435ab9106a4abdb2e3d5c076dc90a9ea9a9fd75e/SCClassLibrary/Common/Control/ResponseDefs.sc#L414

While we could use functions as indices, retrieving them by index (in .value) won't work. Isn't `argTemplate` a more apt location for this matching functionality?

**Comment by jamshark70** [comment]
```
a = { 10.rand };

if(true, { a })  // answers "a Function"

if(true, a)  // answers an integer
```

In this context, if msgNum is a matching function, it must not be evaluated as part of a conditional during initialization. It's valid to execute only when receiving a MIDI message (when there is an incoming datum to match).

Omitting the function braces for the false branch in those two if statements means that a msgNum function will not simply be passed through as a function (correct), but evaluated.

The use of comma syntax here makes it syntactically valid to omit the braces that normally delimit if branches, but it alters the semantics in a way that isn't valid.

**Comment by elgiano** [comment]
This is true, but msgNum is not supposed to be a Function anyway:
> msgNum	
An Integer indicating the MIDI message number (note number, control number, or program number) for this MIDIdef. This can be an array. If nil, the MIDIdef will respond to messages of all possible message numbers.

I think argTemplate is the bugged one here, because it currently allows matching only for the first item of the midi message (i.e. its velocity), while it would be supposed to match the whole message:

https://github.com/supercollider/supercollider/blob/435ab9106a4abdb2e3d5c076dc90a9ea9a9fd75e/SCClassLibrary/Common/Control/ResponseDefs.sc#L1139

This allows argTemplate to be an Integer, but limits it to match velocity only

**Comment by jamshark70** [comment]
Ok, I can agree that msgNum shouldn't be a function, but let's do that explicitly and throw a meaningful error instead of relying on an accident.

I'd say definitely the argTemplate should get other arguments too. E.g., for note on, filtering by velocity would be more rare than filtering by note number, but the rare case is the only one that's supported.

**Comment by elgiano** [comment]
agreed! so we could check before matching, not to break current behavior:
```supercollider
if(argTemplate.isNumber) {
    if(argTemplate.matchItem(testMsg.first), {func.value(*testMsg)});
} { 
    if(argTemplate.matchItem(testMsg), {func.value(*testMsg)});
}
```

**Comment by muellmusik** [comment]
> This is true, but msgNum is not supposed to be a Function anyway:

Yes, I was just going to say the same. I've changed the labels to reflect this.

Your example use case can be accommodated by just providing an array of note numbers for which the function would be true. Something more exotic regarding changing conditions could be accommodated by replacing the MIDIdef when needed. So IIUC there's not a functionality issue here. It was a long time ago, but I think this was a deliberate tradeoff IIRC, as note numbers (and other values for msgNum) are a finite set, so most cases can be straightforwardly and efficiently handled with an array.

> I think argTemplate is the bugged one here

Again not a bug, the doc is quite clear.

One way to accommodate this special case of function rather than defined set would be with a custom dispatcher. The dispatcher argument was included specifically to accommodate special cases.

**Comment by elgiano** [comment]
> Again not a bug, the doc is quite clear.
The doc:
> An optional Integer or Function (or object which responds to the method Methods: matchItem) used to match the value of an incoming MIDI message. (e.g. velocity, control value, program number, etc.). If a Function, it will be evaluated with the message value as an argument, and should return a Boolean indicating whether the message matches and this MIDIdef should respond.

Not 100% clear and accurate for me, since the message's value is an array
 - if argTemplate is an Integer, which element of the array it should match?
 - if argTemplate is a Function, it promises to pass the message as argument, but it actually passes only its first item

I agree with the doc that we should pass the whole message to functions. idk what to do with integers, but we could just keep current behavior for now

**Comment by elgiano** [comment]
I sketched an implementation for MIDIValueMatcher:value. This treats Collection argTemplates like OSCFunc does, adds a new behavior for Function (as described in current helpfile), and keeps current behavior for other cases, even though I feel we should deprecate them (i.e. no single Integer as argTemplates):
```diff
diff --git a/SCClassLibrary/Common/Control/ResponseDefs.sc b/SCClassLibrary/Common/Control/ResponseDefs.sc
index a319aa02b..4d0641935 100644
--- a/SCClassLibrary/Common/Control/ResponseDefs.sc
+++ b/SCClassLibrary/Common/Control/ResponseDefs.sc
@@ -1136,6 +1136,17 @@ MIDIValueMatcher : AbstractMessageMatcher {
        init {|argArgTemplate, argFunc| argTemplate = argArgTemplate; func = argFunc; }
 
        value {|...testMsg|
-               if(argTemplate.matchItem(testMsg.first), {func.value(*testMsg)});
+               if(argTemplate.isFunction) {
+                       if(argTemplate.matchItem(*testMsg), {func.value(*testMsg)});
+                       ^ this;
+               }; 
+               if(argTemplate.isCollection) {
+                       argTemplate.do({|item, i|
+                               if(item.matchItem(testMsg[i]).not, {^this});
+                       });
+                       func.value(*testMsg)
+               } {
+                       if(argTemplate.matchItem(testMsg.first), {func.value(*testMsg)});
+               }
        }
 }
```

Usage:
```
// array template syntax like OSCFunc
// match a single note (not really needed, since it's done better by msgNum)
MIDIdef.noteOn(\test, { |... args| args.postln }, argTemplate: [nil,50]);
// match by function
MIDIdef.noteOn(\test, { |... args| args.postln }, argTemplate: [_>60,_<=49]);

// function syntax
MIDIdef.noteOn(\test, { |... args| args.postln }, argTemplate: {|vel,note|
	(note <= 49).and(vel>60) 
});
```

I think the best way out here would be to deprecate both Integers AND Functions, and just stick with OSCFunc style argTemplate behavior. Then argTemplate would be what its name clames: a template for arguments, a bank of filters for arguments to the function. And we would get a more coherent interface to Responders

**Comment by muellmusik** [comment]
> Not 100% clear and accurate for me, since the message's value is an array

MIDI messages follow a general format of value and num (or num only for some types). The first of these are what is passed to argTemplate, which is what it says. The full message array is not passed there, only to the main func after matching. A further note could be added to clarify if you have a proposal, but FWIW I don't read it the way you do.

> This treats Collection argTemplates like OSCFunc does

No, the analogous behaviour to what you propose would be if OSCArgsMatcher passed the OSC path to the argTemplate. It doesn't, only the additional message arguments. Current OSCFunc is the analogous to the current MIDIFunc behaviour.

> I think the best way out here would be to deprecate both Integers AND Functions, and just stick with OSCFunc style argTemplate behavior.

I would be opposed to deprecating anything, especially what I think are common cases in favour of a rare one. This will break stuff.

> And we would get a more coherent interface to Responders

As I see it it's coherent now. What you're proposing is additional functionality. I'm slightly uncomfortable with the idea as it muddies the waters about what's checked where. It might be worth it, but again this seems rare. Is there a use case that can't be solved with the existing interface?

**Comment by elgiano** [comment]
> FWIW I don't read it the way you do.

I see your point now, it's just confusing IMO that the docs refer to the message's value, and it means "only the first value". I get that velocity is the "value" of a note, like the "value" of a cc message. The confusion is between "what is called value" and "the value of a message" (message is an Array, and Array.value is the array).

> the analogous behaviour to what you propose would be if OSCArgsMatcher passed the OSC path to the argTemplate

I don't really agree. I think MIDIFunc current behavior would be analogous to OSCArgsMatcher passing *only* the path to argTemplate, since the path is the first "value" of that message.

> especially what I think are common cases in favour of a rare one

Is it so common to filter notes by specific velocity, or ccs by specific value (or arrays of them)? 

> As I see it it's coherent now. What you're proposing is additional functionality. I'm slightly uncomfortable with the idea as it muddies the waters about what's checked where. It might be worth it, but again this seems rare. Is there a use case that can't be solved with the existing interface?

Ok. I don't know of one. You can do OP and similar things with full arrays or checks inside the MIDIFunc. Maybe it was just a matter of not wanting to pass an array of possibly hundred items where a single comparison would do. So, maybe not crucial.

As I get it, the current, coherent interface for argTemplate is: 
- You can't filter OSC paths -> You can't filter midi notes, cc nums and so
- you can filter only the message value: for OSC messages that is all the arguments, for MIDI it's the first one only
 
So, all in all:
- a different argTemplate behavior for MIDIFunc would be a new feature
- the current behavior needs at least a disambiguation in the docs, to avoid that confusion

**Comment by jamshark70** [comment]
I see -- multiple misunderstandings on my part.

I hadn't realized that the MIDI response system installs multiple references to the same function proxy. I was trying the function to avoid an O(n) array scan for each incoming message, but now I see that it was never doing that.

Tbh I'd be happy to close this, or revisit the documentation (to be clear that it isn't using matchItem, and to explain more explicitly that argTemplate is not an analog to the OSCFunc argTemplate).

**Comment by muellmusik** [comment]
> The confusion is between "what is called value" and "the value of a message" (message is an Array, and Array.value is the array).

Yes, could be clearer for sure!



> I don't really agree. I think MIDIFunc current behavior would be analogous to OSCArgsMatcher passing _only_ the path to argTemplate, since the path is the first "value" of that message.

Well, path is path, arguments are arguments. Path is checked top of the chain, and there's a very clearly defined spec in OSC for path matching, etc. which is handled. It's perhaps true that there is an element of arbitrariness in the semantics of course, but it makes sense to me.



> Is it so common to filter notes by specific velocity, or ccs by specific value (or arrays of them)?

No. But the common case is handled by the msgNum arg. argTemplate and custom dispatchers are there to allow for more exotic cases. That was the design decision 9 years ago. Perhaps it was the wrong one, but...



> * You can't filter OSC paths -> You can't filter midi notes, cc nums and so
> * you can filter only the message value: for OSC messages that is all the arguments, for MIDI it's the first one only

Um no. You can 'filter' if I understand correctly, with the msgNum arg and the path arg for OSC. As I said I'm not sure of a use case that can't be handled.

> I hadn't realized that the MIDI response system installs multiple references to the same function proxy. I was trying the function to avoid an O(n) array scan for each incoming message, but now I see that it was never doing that.

Speed was a major design criterion when these were done. Perhaps even too much so. But it does do extra work to make dispatching fast. This is less of an issue with MIDI than with OSC, where you would be more likely to have hundreds of messages per second or more, but as they share a lot of code of design it seemed no reason not to do it.



> Tbh I'd be happy to close this, or revisit the documentation (to be clear that it isn't using matchItem, and to explain more explicitly that argTemplate is not an analog to the OSCFunc argTemplate).

To me it seems analogous as I think in this case of note number as being like path (as MIDI message type seems more fundamental, and is handled in the constructor anyway) but again I can see the confusion. More clarity is always good!

**Comment by elgiano** [comment]
Thanks for the explanations! It makes more sense now for me to accept the current behavior.
I tried to sketch an addition to the docs, but I can't help falling back to the thought that `argTemplate` means a template to the arguments passed to func (val, num, chan and src), even if it says that it matches only val, and val is defined in the description of func. Would you have any idea for clarifying this? Or maybe an example would be better?

**Comment by muellmusik** [comment]
Thanks @elgiano for looking at this. Looking at the existing doc, it does try to make this clear with "(e.g. velocity, control value, program number, etc.)". Perhaps it would help to specify what it doesn't do? How about this:

An optional Integer or Function (or object which responds to the method Methods: matchItem) used to match the _value_ of an incoming MIDI message. _Value_ in this case refers to velocity, control value, program number, etc., depending on message type. Not that msgNum is not passed here, as it is checked using the msgNum argument. If a Function, it will be evaluated with the message value as an argument, and should return a Boolean indicating whether the message matches and this MIDIFunc should respond.

An example would be great and help a lot I think. Perhaps...

```
// respond only to high velocity notes.
MIDIdef.noteOn(\highVelocitiesOnly, {arg ...msg; msg.postln; "High velocity note!".postln}, argTemplate: {|velocity| velocity > 89 });
```

**Comment by elgiano** [comment]
Thanks @muellmusik  for finding the time to rephrase it!
By looking at your proposal, I think the confusion here is not so much about msgNum, but about the fact that argTemplate is about only the first "arg" (val) and not about everything that func receives. Maybe more in this direction?

An optional Integer or Function (or object which responds to the method Methods: matchItem) used to match the value of an incoming MIDI message. Value in this case refers only to the first argument passed to func (i.e. velocity, or control value, or program number, etc., depending on message type), and not to an array of all the others. If a Function, it will be evaluated with the message value as an argument, and should return a Boolean indicating whether the message matches and this MIDIFunc should respond.

The example is clear an helpful!

**Comment by muellmusik** [comment]
> By looking at your proposal, I think the confusion here is not so much about msgNum, but about the fact that argTemplate is about only the first "arg" (val) and not about everything that func receives. Maybe more in this direction?

Yes, it probably should have been renamed. It was copied from OSCFunc IIRC.

How about using the term val for consistency with the rest of the doc and to distinguish with value (confusing in an SC context), and then clarifying further by exclusion.

An optional Integer or Function (or object which responds to the method Methods: matchItem) used to match the MIDI _val_ of an incoming MIDI message. _Val_ in this case refers to velocity, control value, program number, etc., depending on message type (see func above for more detail). Note that num, chan, and source are not passed here, as they are checked using their corresponding arguments (see above). If a Function, it will be evaluated with the message value as an argument, and should return a Boolean indicating whether the message matches and this MIDIFunc should respond.

I think that's clear to the point of being pedantic. :-)

**Comment by elgiano** [comment]
@muellmusik 
thanks for this, I think it's clearer now! I like that it points to the other arguments, which are usable to check for num, chan and src, so that one can find a solution even if it's not within argTemplate :)
